feat: return job IDs from Worker::perform_later() for status tracking

NewtTheWolf opened the pull request:
## Summary

This PR implements job ID returns from `Worker::perform_later()` to enable job status tracking, addressing #1623.

## Changes

- Modified `Queue::enqueue()` to return `Result<Option<String>>` instead of `Result<()>`
- Modified `BackgroundWorker::perform_later()` to return `Result<Option<String>>` instead of `Result<()>`
- All queue providers (Redis, PostgreSQL, SQLite) now return their internally tracked job IDs
- Returns `Some(job_id)` when using BackgroundQueue mode with a configured provider
- Returns `None` for ForegroundBlocking, BackgroundAsync modes, or BackgroundQueue without a provider

## Testing

- Updated all existing `enqueue` tests to verify job ID returns
- Added validation in tests to verify that returned job IDs are valid ULIDs
- Added new test `queue_enqueue_returns_job_id` to verify behavior for different queue modes
- All bgworker tests pass successfully with `cargo test --features bg_redis,bg_pg,bg_sqlt`

## Documentation

- Updated API documentation for `Queue::enqueue()` method
- Updated user documentation in `docs-site/content/docs/processing/workers.md`
- Added usage example showing how to use the returned job ID
- Added changelog entry

## Notes for Reviewers

- This is a non-breaking change - existing code that ignores the return value continues to work
- The implementation follows the approach suggested in the original issue discussion

Implements #1623

NewtTheWolf committed:
feat: return job IDs from Worker::perform_later() for status tracking (#1623)

NewtTheWolf committed:
docs: add changelog entry for job ID feature

NewtTheWolf committed:
fix: add backticks to documentation for clippy

kaplanelad committed:
Merge branch 'master' into feature/expose-worker-job-ids

jondot committed:
Merge branch 'master' into feature/expose-worker-job-ids

kaplanelad committed:
Merge branch 'master' into feature/expose-worker-job-ids

NewtTheWolf committed:
Merge branch 'master' into feature/expose-worker-job-ids

kaplanelad commented:
Hey @NewtTheWolf,
I tested this and I think there’s an issue.
### Steps to Reproduce

1. **Create a clean project:**
   ```
   ❯ loco new
   ✔ ❯ App name? · myapp
   ✔ ❯ What would you like to build? · Saas App with server side rendering
   ✔ ❯ Select a DB Provider · Sqlite
   ✔ ❯ Select your background worker type · Async (in-process tokio async tasks)
   ```
2. **Generate a worker:**
   ```
   cargo loco generate worker test
   ```

3. **Generate a test controller:**
   ```
   cargo loco generate controller test --api
   ```
4. **Add the following code to the controller:**
   ```
   #[debug_handler]
   pub async fn index(State(ctx): State<AppContext>) -> Result<Response> {
    let job: Option<String> =
        workers::test::Worker::perform_later(&ctx, workers::test::WorkerArgs {}).await?;
    println!("job: {job:#?}",);

    format::empty()
   }
   ```
5. **Run the server with the worker:**
   ```
   cargo loco start --server-and-worker
   ```
6. **Open the endpoint in your browser:** http://localhost:5150/api/tests

***Expected***
I should see a job ID in the logs.

Actual i got: job: None

NewtTheWolf committed:
Merge branch 'master' into feature/expose-worker-job-ids

NewtTheWolf commented:
it would make sense absoloutly. i generally like the idea of https://github.com/loco-rs/loco/pull/1624#discussion_r2520711999

sorry i really didnt had time for this, my goal is to update this pr this week!

NewtTheWolf committed:
Merge branch 'master' into feature/expose-worker-job-ids

mccormickt commented on src/bgworker/mod.rs:
I believe so. It would be nice to be able to read the queue from other parts of the application - like an API endpoint that returns job details.

NewtTheWolf replied:
Done

NewtTheWolf commented:
hey @mccormickt, sorry had a loot going on in my live lately. But your Message was a small boost to my Motivation! i will start now! first of all making the Tokio Worker to return an ID so it is no longer Optional and then making the fn `get_jobs` public!

NewtTheWolf committed:
Merge branch 'master' into feature/expose-worker-job-ids

NewtTheWolf committed:
feat: expose tokio task id in BackgroundAsync worker mode

NewtTheWolf committed:
feat: make Queue::get_jobs public for external job querying

NewtTheWolf committed:
feat: always return job ID from perform_later

NewtTheWolf committed:
Merge branch 'master' into feature/expose-worker-job-ids

NewtTheWolf committed:
fix: add missing errors doc for `Queue::get_jobs`

mccormickt reviewed (approved):
LGTM! Thanks for hanging in there and getting this across the line.
cc @kaplanelad

NewtTheWolf commented:
@kaplanelad

i will take a look into it.
> BackgroundAsync — returns a tokio::task::Id (an integer like "42"), not a ULID and not in any queue. This is the most concerning case — if a user switches from BackgroundQueue to BackgroundAsync, their code that passes the returned ID to get_jobs() will silently fail. The format is also completely different (integer vs ULID).

Yes! but i think we should do an internal mapping from Ulid to TaskId in case we have a usecase for it in the future like with a Map (or smth similar)

> Previously, if BackgroundQueue mode was configured but no provider was set, it logged an error and continued. Now it returns Err(...). This will crash apps that previously worked in a degraded state. Could we keep this as a warning/error log, or at minimum call this out as an intentional behavior change?

i think so yes!

it is on my todo board!

NewtTheWolf committed:
feat: consistent ULID job IDs across all worker modes

NewtTheWolf committed:
Merge branch 'master' into feature/expose-worker-job-ids

NewtTheWolf committed:
fix: correct import ordering for cargo fmt

NewtTheWolf commented:
Hey @kaplanelad, I addressed your points:

Inconsistent ID semantics:
BackgroundAsync now generates a ULID instead of returning the tokio task::Id. All three modes return consistent
ULID-formatted IDs now. I also added an AsyncJobTaskMap that internally maps the ULID to the tokio task ID, so we
retain access to it if needed in the future. The map is accessible via the SharedStore.

Error on missing provider:
Reverted back to tracing::error!() with a fallback ULID instead of returning Err(...), so apps in a degraded state
won't crash.

NewtTheWolf committed:
fix: add missing must_use attribute on AsyncJobTaskMap::remove

jondot commented:
Thanks @NewtTheWolf — `perform_later` returns the job id in 1.0.0, exactly what this set out to do (alongside `perform_later_with_priority`). Your PR drove it; closing in favor of the 1.0.0 implementation. Much appreciated — see the CHANGELOG for the full background-jobs changes.
